### PR TITLE
Bugfix/1582/rewrite toggle chat etc

### LINKED ISSUE #1582: rewrite "toggle Chat" feature (fixes many bugs)
The "toggle chat" feature (https://github.com/nextcloud/talk-android/pull/912) obviously caused a lot of bugs.
The main problem is that multiple instances of the ChatController are created here:
https://github.com/nextcloud/talk-android/blob/master/app/src/main/java/com/nextcloud/talk/activities/MagicCallActivity.kt#L92

So inside the ChatController an intent is sent to start the MagicCallActivity
https://github.com/nextcloud/talk-android/blob/master/app/src/main/java/com/nextcloud/talk/controllers/ChatController.kt#L2137

and in the MagicCallActivity another ChatController is created. As a result there is a mess with different controllers and sessions.

The known bugs that are caused by this feature should be at least:
- https://github.com/nextcloud/talk-android/issues/1232
- accept the call right after granting permissions doesn't work
   - might fix https://github.com/nextcloud/talk-android/issues/1309
- https://github.com/nextcloud/talk-android/issues/935
- https://github.com/nextcloud/talk-android/issues/938
- https://github.com/nextcloud/talk-android/issues/1584
- https://github.com/nextcloud/talk-android/issues/1583
- i guess that "one-way" connections where only one person could hear the other are caused by the code of the toggle chat feature

A clean solution might be only possible when conductor is replaced by native android solutions (see https://github.com/nextcloud/talk-android/issues/1551).

### PR DESCRIPTION
**Pull request by mahibi**
resolves #1582  by picture-in-picture support

call screen | launcher | chat view pip | chat view pip large 
------- | -------- | -------- | -------- 
grafik | grafik | grafik | grafik

### CODE REVIEW

File: app/src/main/java/com/nextcloud/talk/activities/CallNotificationActivity.java [hunk 1]
```diff
@@ -517,4 +485,65 @@ private void playRingtoneSound() {
...
+        binding.incomingCallRelativeLayout.setVisibility(View.INVISIBLE);
+    }
+
+    public void updateUiForNormalMode(){
+        binding.callAnswerButtons.setVisibility(View.VISIBLE);
+        binding.incomingCallRelativeLayout.setVisibility(View.VISIBLE);
+    }
+
+    // TODO: dismiss keyguard works, but whenever accepting the call and switch to CallActivity by intent, the
+    //  lockscreen is shown (although CallActivity also dismisses the keyguard in the same way.)
```
[the remark this answers is not in the document]
Comment (mahibi) [reply app/src/main/java/com/nextcloud/talk/activities/CallNotificationActivity.java:535]: postponed. created issue: https://github.com/nextcloud/talk-android/issues/1667

**Comment by mahibi** [comment]
please review whenever your time allows it @timkrueger @nickvergessen @AndyScherzinger

**Comment by andyxheli** [comment]
@mahibi I tested the QA out and seem to work really good with the PIP . Only thing I seen is that the picture is super zoomed on IOS platform you're able to double tap the screen and zooms out the resolution

**Comment by nickvergessen** [comment]
Works good, but the video seems to freeze when tabbing the PiP so the controls (like open in fullscreen again) show

**Comment by nickvergessen** [comment]
Also I have a setting in the PiP where I can disable PiP mode. If I do that, the button should be hidden (in case you can even detect it=

**Comment by timkrueger** [comment]
With a touch on the PiP  window it increases a little bit. But during that the front camera stream is shown and it stays there also if the size of the PiP windows decreases.

https://user-images.githubusercontent.com/606665/140519038-35ee1aba-15e2-4b34-8dca-ebdf1652524b.mp4

**Review by timkrueger — APPROVED** [review]
:1st_place_medal:

**Comment by mahibi** [comment]
@andyxheli thanks for testing :+1: 

> Only thing I seen is that the picture is super zoomed

can't reproduce this. does it only happen with this branch or also for other versions?

> also the button are a little cut out on the side

this is now fixed

**Comment by andyxheli** [comment]
> @andyxheli thanks for testing 👍
> 
> > Only thing I seen is that the picture is super zoomed
> 
> can't reproduce this. does it only happen with this branch or also for other versions?
> 
> > also the button are a little cut out on the side
> 
> this is now fixed

You're Welcome! Thank you for everything. 

 Looks like it happens with others versions also. For example if you have someone join the call with a 1080p camera or a user that turned their phone landscape the app will cut off and not show the whole image normally on the ios platform you can doble tap and it will show a wide screen view

**Comment by andyxheli** [comment]
@mahibi Found some issues 

1. The app crashed when you're in a chat with multiple users when you go into PIP mode and back no error just closes out

2. When on a call with multiple users and you rotate the phone screen to landscape the phone camera froze " on the phone end you dose see it front but other users see you frozen" 

3.When  When on a call with multiple users and you rotate the phone screen to landscape the participants video feed starts to flaker on the phone.

**Comment by mahibi** [comment]
> Also I have a setting in the PiP where I can disable PiP mode. If I do that, the button should be hidden (in case you can even detect it=

button is now hidden when PiP is disabled in android settings. Also the navigation bar (home & back button) is now hidden for this case.

**Comment by mahibi** [comment]
> Works good, but the video seems to freeze when tabbing the PiP so the controls (like open in fullscreen again) show

as discussed, this was the same bug that tim reported. So the selfvideo was frozen, but this should now be hidden on all devices when you tap the pip view.

**Comment by andyxheli** [comment]
Check out the video :) Hope this helps
https://user-images.githubusercontent.com/59488153/141140734-5e97e1c7-a020-4fad-aa34-3ebcc172bf57.mp4

**Comment by mahibi** [comment]
>     5. Let me know I created a screen recorded let me know if you have an email I can send a shared link too.

@andyxheli thank you, please send the mail to dev@mhibbe.de

**Comment by andyxheli** [comment]
@mahibi Link has been sent please let me know if you got it

**Comment by mahibi** [comment]
>     .....
>     5. Let me know I created a screen recorded let me know if you have an email I can send a shared link too.

thanks @andyxheli got the mail with the video. will try to take a look in the next days!

**Comment by andyxheli** [comment]
> > ```
> > .....
> > 5. Let me know I created a screen recorded let me know if you have an email I can send a shared link too.
> > ```
> 
> thanks @andyxheli got the mail with the video. will try to take a look in the next days!

You're Welcome! Thank you. Let me know if you need need anything else. Happy to help.

File: app/src/main/res/values/strings.xml [hunk 2]
```diff
@@ -221,6 +221,10 @@
     <string name="nc_call_state_with_phone">%1$s with phone</string>
     <string name="nc_call_state_with_video">%1$s with video</string>
 
+    <!-- Picture in Picture -->
+    <string name="nc_pip_microphone_mute">mute microphone</string>
+    <string name="nc_pip_microphone_unmute">enable microphone</string>
```
Comment (AndyScherzinger) [inline app/src/main/res/values/strings.xml:226]: I would guess we do sentence casing here, so
`Mute microphone`
`Enable microphone`

**Comment by AndyScherzinger** [comment]
Looks good code-wise, didn't test.

**Comment by mahibi** [comment]
thanks everyone for testing!!

i merged this branch because as far as i see the bugs that are reported here..
- already existed before and have nothing to do with the changes in this branch
- can't be reproduced yet from my side
in any case, please create issues with detailed information.

**Comment by andyxheli** [comment]
@mahibi Do i need to report a new issue i turned off PIP and now the app just crashes cant join a call ?

**Comment by mahibi** [comment]
> @mahibi Do i need to report a new issue i turned off PIP and now the app just crashes cant join a call ? If i turn it back on app starts to work again
> 21-11-11-10-08-13.mp4

ok good to know, yes please create an issue for that. including which OS and smartphone this is, please :+1:

**Pull request merged**